source-container-build/app/test_source_build.py

```diff
@@ -0,0 +1,1154 @@
+                completed_proc.stdout = "https://githost/org/app.git"
+            else:
+                completed_proc.stdout = ""  # No other calls depend on the stdout
+            return completed_proc
+
+        run.side_effect = run_side_effect
+
+        cli_cmd = [
+            "source_build.py",
+            "--bsi",
+            self.bsi,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+        ]
+        with patch("sys.argv", cli_cmd):
+            with self.assertLogs(source_build.logger, level=logging.DEBUG) as logs:
+                rc = source_build.main()
+
+        self.assertEqual(0, rc)
+
+        for log in logs.output:
+            if "workspace directory " in log:
+                self.assertRegex(log, r"workspace directory .+-source-build-workspace")
+                break
+        else:
+            self.fail("Cannot find log line logging created temporary workspace directory.")
+
+    def test_bsi_script_has_to_exist(self):
+        fd, bsi = mkstemp()
+        os.close(fd)
+        os.unlink(bsi)
+
+        cli_cmd = [
+            "source_build.py",
+            "--workspace",
+            self.work_dir,
+            "--bsi",
+            bsi,
+            "--write-result-to",
+            self.result_file,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+        ]
+        with patch("sys.argv", cli_cmd):
+            with self.assertRaises(SystemExit):
+                source_build.main()
+
+    def test_bsi_script_has_to_be_executable(self):
+        fd, bsi = mkstemp()
+        os.close(fd)
+
+        def _remove_bsi():
+            os.unlink(bsi)
+
+        self.addCleanup(_remove_bsi)
+
+        cli_cmd = [
+            "source_build.py",
+            "--workspace",
+            self.work_dir,
+            "--bsi",
+            bsi,
+            "--write-result-to",
+            self.result_file,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+        ]
+        with patch("sys.argv", cli_cmd):
+            with self.assertRaises(SystemExit):
+                source_build.main()
+
+    def test_specified_work_dir_has_to_exist(self):
+        work_dir = mkdtemp()
+        os.rmdir(work_dir)
+
+        cli_cmd = [
+            "source_build.py",
+            "--workspace",
+            work_dir,
+            "--bsi",
+            self.bsi,
+            "--write-result-to",
+            self.result_file,
+            "--source-dir",
+            self.app_source_dirs.cloned_dir,
+            "--output-binary-image",
+            OUTPUT_BINARY_IMAGE,
+        ]
+        with patch("sys.argv", cli_cmd):
+            with self.assertRaises(SystemExit):
+                source_build.main()
```

Comment:
We have a lot of tests for failure cases, but they don't assert any specific exception type or error message. Do we even have error handling for these cases? Is it necessary to test them?

Reply:
Two kinds of failure cases.

One of them happens in the individual method called during the build process, for example `extract_blob_member`. Such methods are not responsible for handling errors, instead, the error will be propagated then be handled by `main`. So, the related tests ensure that failures, the failed subprocess run, are raised as a subprocess error.

Another kind of failure is caused by input, like what your comment is for. It is handled by the `add_argument` `type` argument. For this case, just ensure the command exits by raising `SystemExit`.

Reply:
Ok, seems reasonable